### PR TITLE
475/adds ckan dataportal

### LINKED ISSUE #475: Integration with CKAN as a data portal
# Overview

An important step for Frictionless Framework is to provide an ability to read and write packages from different data portals (CKAN/Github/Zenodod/etc) so the users can publish and access their packages easily and using a straightforward API. This issue is for CKAN integration. The implementation is already prototyped in v5 branch.

### Read package

Read package from a dataset that has a datapackage.json/yaml:

```python
package = Package("https://dados.gov.br/dataset/consumo-de-energia-orgaos-publicos")
package = Package.from_ckan(...) # alias
```

Read package from a repository without a datapackage.json/yaml. We probably need to filter files and add only CSV/XLS(X) to the package. Also CkanControl should have this configurable. We need to map as much as possible metadata provided by CKAN (for now, using `frictionless-ckan-mapper`):

```python
package = Package("https://demo.ckan.org/dataset/sample-dataset-1")
package = Package("<link>", control=portals.CkanControl(formats=['csv']))
```

### Write package

Publish a package on CKAN (for now, only if the dataset doesn't exist). Also we need to provide an ability to store credentials in ENV/etc. We need to map as much as possible metadata provided by Package:

```python
package.to_ckan()
```

### Read catalog

Read catalog from ckan search. Design some search configurations like `limit` and `offset` (pagination).

```python
catalog = Catalog(control=portals.CkanControl(baseurl=<ckan-instance-url>, search="<frictionless>")
for package in catalog.packages:
  print(package.name)
```

## Plan

- [ ] prototype the functionality based on the functional requirements
- [ ] get feedback from @roll on the implementation
- [ ] finish the implementation
- [ ] design the testing approach (probably using pytest.vcr fro reading/from what instance? but how to test writing?) (sync with Adria on the best practices of testing against CKAN instances)
- [ ] write a great deal of tests to be sure that the integration works correctly
- [ ] write a comprehensive tutorial - https://framework.frictionlessdata.io/docs/tutorials/tutorials-overview (new section Portals Tutorials)

### PR DESCRIPTION
**Pull request by aivuk**
- fixes #475

This PR adds support to CKAN  as a data portal.

 - [x] Reading individual packages  
 - [x] Publishing individual packages to a CKAN instance
 - [x] Fetching a list of packages using CKAN search
 - [x] Documentation
 - [ ] Local tests

### CODE REVIEW

**Comment by roll** [comment]
Great @aivuk!

Can you please resolve the conflicts as some API changes have been introduced lately (Manager -> Adapter)?

@shashigharti Can you please test and review as I will not be available?

**Comment by aivuk** [comment]
@roll I fixed the conflicts but I did need to rebase this branch on main and after that force-pushed the branch again.

File: docs/portals/ckan.md [hunk 1]
```diff
@@ -1,5 +1,200 @@
...
+
+You can see in the example above that 1000 packages were download from a total
+13786 packages. You can download other packages passing an offset as:
+
+```python tabs=Python
+
+import frictionless
+from frictionless import portals, Catalog
+    
+ckan_control = portals.CkanControl(baseurl='https://dados.gov.br', ignore_package_erros=True, results_offset=1000)
```
Comment (shashigharti) [inline docs/portals/ckan.md]: type in param name : ignore_package_erros

**Comment by shashigharti** [comment]
@roll @aivuk everything looks great! 

Added a small comment related to typo. And also this URL ("https://dados.gov.br") in example  was not working.

File: docs/portals/ckan.md (hunk 1 above)
Comment (aivuk) [reply docs/portals/ckan.md]: fixed, thanks

**Comment by aivuk** [comment]
> @roll @aivuk everything looks great!
> 
> Added a small comment related to typo. And also this URL ("https://dados.gov.br") in example was not working.

I changed to legado.dados.gov.br that is working but I don't know until when.

**Review by shashigharti — APPROVED** [review]
Thanks!

**Pull request merged**